README.md
# ln-constr-parser

[![NPM version](https://img.shields.io/npm/v/@synonymdev/ln-constr-parser?color=%23FFAE00&style=flat-square)](https://www.npmjs.com/package/@synonymdev/ln-constr-parser)

Convenient, dependency-free Lightning connection string parser for front- and backend.

## Usage

```typescript
import {parseConnectionString} from '@synonymdev/ln-constr-parser';

const connectionString = '0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f@127.0.0.1:9735';
const {host, hostType, port, pubKey} = parseConnectionString(connectionString);

console.log(host); // 127.0.0.1
console.log(hostType); // ipv4. Depending on the address provided, it could also be ipv6, torv3 or domain.
console.log(port); // 9735
console.log(pubKey); // 0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f
```

The `port` is optional. You can make the port mandatory so the parser will throw an error. This also avoids the
ipv6 ambiguity problem (see below).

```typescript
parseConnectionString(connectionString, {portMandatory: true})
```

### Parse Errors

In case of an invalid connection string, the parser will throw a detailed error of what is wrong.

```typescript
import {parseConnectionString, ParseFailureCode, ParseError} from '@synonymdev/ln-constr-parser';

try {
    const connectionString = 'pubkey@host:port';
    parseConnectionString(connectionString);
} catch (e) {
    if (e instanceof ParseError) {
        console.log('Parse failed. Reason:', e.code)
        if (e.code === ParseFailureCode.INVALID_HOST) {
            console.log('Host value is invalid.')
        } else if (e.codee === ParseFailureCode.INVALID_IPV6) {
            console.log('IPv6 host is invalid.')
        } else if (e.codee === ParseFailureCode.AMBIGUOUS_IPV6) {
            console.log('IPv6 host is ambiguous. Use square brackets [] like pubkey@[ipv6]:port.')
        } else if (e.code === ParseFailureCode.INVALID_PORT) {
            console.log('Invalid port.')
        } else if (e.code === ParseFailureCode.INVALID_PUBKEY) {
            console.log('Invalid lightning pubkey.')
        } else if (e.code === ParseFailureCode.INVALID_ATS) {
            console.log('The connection string must contain one single @ symbol.')
        }
    }
}
```

## Supported Address Formats

The library supports all commonly used address formats in the form of 
- `pubkey@host:port`
- `pubkey@host`
- `pubkey@[host]:port`
- `pubkey@[host]`

**ipv4**
- Regular `pubkey@127.0.0.1:port`.
- No port `pubkey@127.0.0.1`.
- Square brackets `pubkey@[127.0.0.1]:port`.
- Square brackets no port `pubkey@[127.0.0.1]`.

**ipv6**
- Square brackets `pubkey@[2001:db8:3333:4444:5555:6666:7777:8888]:port`.
- Square brackets compressed `pubkey@[2001:db8::8888]:port`.
- Square brackets no port `pubkey@[2001:db8:3333:4444:5555:6666:7777:8888]`.
- Square brackets compressed no port `pubkey@[2001:db8::8888]`.
- Regular uncompressed `pubkey@2001:db8:3333:4444:5555:6666:7777:8888:port`.
- No port uncompressed `pubkey@2001:db8:3333:4444:5555:6666:7777:8888`.

> **⚠️ Ambiguity problem** If you don't make the port mandatory, it is adviced to use square brackets `[]` with IPv6. It is impossible to separate a compressed IPv6 from the port. Example:
> `2001:db8::8888:9735` may as well be a IPv6 (`2001:db8::8888:9735`) OR a IPv6:port (`2001:db8::8888` and port `9735`).
> The library will do its best to parse it correctly but will throw an error in case of ambiguity.

**torv3**
- Regular `pubkey@giexynrrloc2fewstcybenljdksidtglfydecbellzkl63din6w73eid.onion:port`.
- No port `pubkey@giexynrrloc2fewstcybenljdksidtglfydecbellzkl63din6w73eid.onion`.
- Square brackets `pubkey@[giexynrrloc2fewstcybenljdksidtglfydecbellzkl63din6w73eid.onion]:port`.
- Square brackets no port `pubkey@[giexynrrloc2fewstcybenljdksidtglfydecbellzkl63din6w73eid.onion]`.

**domain**
- Regular `pubkey@synonym.to:port`.
- No port `pubkey@synonym.to`.
- Square brackets `pubkey@[synonym.to]:port`.
- Square brackets no port `pubkey@[synonym.to]`.




## Parse Subcomponents

The library also provides functions to parse the individual components of a connection string.

```typescript
import {parseAddress, parseHost, parsePort, parsePubkey} from '@synonymdev/ln-constr-parser';

const {host, hostType, port} = parseAddress('127.0.0.1:9000')
const {host, hostType} = parseHost('127.0.0.1')
const port = parsePort('900')
const pubkey = parsePubkey('0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f')
```


---

May the ⚡ be with you.

CHANGELOG.md
# Changelog

## 1.0.1 - 26.06.2013 - Severin Bühler

- (feat) Added `portMandatory` parameter to `parseConnectionString` function to avoid the ipv6 ambiguity problem.
- (feat) Added distinct `ParseFailureCode.AMBIGUOUS_IPV6` error in case the provided ipv6 address can't be correctly separated.

## 1.0.0 - 16.06.2013 - Severin Bühler

- (feat) first version.


src/parse.spec.ts
import { ParseError, ParseFailureCode } from './ParseError';
import {parseConnectionString, splitHostAndPort, parseHost, parseAddress, parsePort, parsePubkey} from './parse';


describe('parse', () => {
    test('splitHostAndPort without []', async () => {
        const {notValidatedHost: host1, notValidatedPort: port1} = splitHostAndPort('127.0.0.1:300')
        expect(host1).toEqual('127.0.0.1')
        expect(port1).toEqual('300')

        const {notValidatedHost: host2, notValidatedPort: port2} = splitHostAndPort('gwdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqd.onion:300')
        expect(host2).toEqual('gwdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqd.onion')
        expect(port2).toEqual('300')

        const {notValidatedHost: host3, notValidatedPort: port3} = splitHostAndPort('127.0.0.1')
        expect(host3).toEqual('127.0.0.1')
        expect(port3).toEqual(undefined)
    });

    test('splitHostAndPort with []', async () => {
        const {notValidatedHost: host1, notValidatedPort: port1} = splitHostAndPort('[127.0.0.1]:300')
        expect(host1).toEqual('127.0.0.1')
        expect(port1).toEqual('300')

        const {notValidatedHost: host2, notValidatedPort: port2} = splitHostAndPort('[2001:db8:3333:4444:5555:6666:7777:8888]:300')
        expect(host2).toEqual('2001:db8:3333:4444:5555:6666:7777:8888')
        expect(port2).toEqual('300')

        const {notValidatedHost: host3, notValidatedPort: port3} = splitHostAndPort('[gwdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqd.onion]:300')
        expect(host3).toEqual('gwdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqd.onion')
        expect(port3).toEqual('300')
    });

    test('splitHostAndPort ipv6 without []', async () => {
        const {notValidatedHost: host1, notValidatedPort: port1} = splitHostAndPort('2001:db8:3333:4444:5555:6666:7777:8888:300')
        expect(host1).toEqual('2001:db8:3333:4444:5555:6666:7777:8888')
        expect(port1).toEqual('300')

        const {notValidatedHost: host2, notValidatedPort: port2} = splitHostAndPort('2001:db8:3333:4444:5555:6666:7777:8888')
        expect(host2).toEqual('2001:db8:3333:4444:5555:6666:7777:8888')
        expect(port2).toBeUndefined()

        try {
            splitHostAndPort('2001:db8:3333:4444:5555::8888')
            expect(true).toEqual(false)
        } catch (e: any) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.AMBIGUOUS_IPV6)
        }

        const {notValidatedHost: host3, notValidatedPort: port3} = splitHostAndPort('2001:db8:3333:4444::7777:8888:300', {portMandatory: true})
        expect(host3).toEqual('2001:db8:3333:4444::7777:8888')
        expect(port3).toEqual('300')
    });

    test('parseHost', async () => {
        const {host: host1, hostType: type1} = parseHost('127.0.0.1')
        expect(host1).toEqual('127.0.0.1')
        expect(type1).toEqual('ipv4')

        const {host: host2, hostType: type2} = parseHost('2001:db8:3333:4444:5555:6666:7777:8888')
        expect(host2).toEqual('2001:db8:3333:4444:5555:6666:7777:8888')
        expect(type2).toEqual('ipv6')

        const {host: host3, hostType: type3} = parseHost('gwdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqd.onion')
        expect(host3).toEqual('gwdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqd.onion')
        expect(type3).toEqual('torv3')

        const {host: host4, hostType: type4} = parseHost('swiss-reign.ch')
        expect(host4).toEqual('swiss-reign.ch')
        expect(type4).toEqual('domain')

        try {
            parseHost('127.0.0.1:')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_HOST)
        }
        try {
            parseHost('127.0.0.1:-1')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_HOST)
        }

        try {
            parseHost(undefined as any)
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_HOST)
        }
    });

    test('parseAddress', async () => {
        const {host: host1, hostType: type1, port: port1} = parseAddress('127.0.0.1:300')
        expect(host1).toEqual('127.0.0.1')
        expect(type1).toEqual('ipv4')
        expect(port1).toEqual(300)

        const {host: host2, hostType: type2, port: port2} = parseAddress('2001:db8:3333:4444:5555:6666:7777:8888:300')
        expect(host2).toEqual('2001:db8:3333:4444:5555:6666:7777:8888')
        expect(type2).toEqual('ipv6')
        expect(port2).toEqual(300)


        try {
            parseAddress('127.0.0.1', {portMandatory: true})
            expect(true).toEqual(false)
        } catch (e: any) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_PORT)
        }
    });

    test('parsePort', async () => {
        expect(parsePort('300')).toEqual(300)
        expect(parsePort(undefined)).toEqual(undefined)
        expect(parsePort('65535')).toEqual(65535)

        try {
            parsePort('0')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_PORT)
        }

        try {
            parsePort('65536')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_PORT)
        }

        try {
            parsePort(undefined, {portMandatory: true})
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_PORT)
        }

        try {
            parsePort('-1')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_PORT)
        }
    });

    test('parsePubkey', async () => {
        expect(parsePubkey('0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f')).toEqual('0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f')
        expect(parsePubkey('0300000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f')).toEqual('0300000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f')

        try {
            parsePubkey('0300000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_PUBKEY)
        }

        try {
            parsePubkey('0300000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5fa')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_PUBKEY)
        }

        try {
            parsePubkey('0100000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5a')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_PUBKEY)
        }
    });

    test('parseConnectionString', async () => {
        const {host: host1, pubkey: pubkey1, port: port1} = parseConnectionString('0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f@127.0.0.1:9000')
        expect(host1).toEqual('127.0.0.1')
        expect(pubkey1).toEqual('0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f')
        expect(port1).toEqual(9000)

        try {
            parseConnectionString('0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f127.0.0.1:9000')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_ATS)
        }

        try {
            parseConnectionString('0200000000a3e@ff613189ca6c4070c89206ad658e286751eca1f29262948247a5f@127.0.0.1:9000')
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_ATS)
        }

        try {
            parseConnectionString('0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f@127.0.0.1', {portMandatory: true})
            expect(true).toEqual(false)
        } catch (e) {
            if (!(e instanceof ParseError)) {
                throw e
            }
            expect(e.code).toEqual(ParseFailureCode.INVALID_PORT)
        }

        expect(() => parseConnectionString('@')).toThrow(ParseError)
        expect(() => parseConnectionString('0200000000a3eff613189ca6c4070c89206ad658e286751eca1f29262948247a5f@')).toThrow(ParseError)

    });

});




src/regex.spec.ts
import * as regex from './regex';


describe('connectionString regex', () => {

    test('ipv4', async () => {
        const ipv4Regex = regex.ipv4Regex

        const match = ipv4Regex.exec('34.65.85.39')
        expect(match).toBeTruthy()

        const match2 = ipv4Regex.exec('34.65.85.3a')
        expect(match2).toBeFalsy()

        const match3 = ipv4Regex.exec('34.65.85.333')
        expect(match3).toBeFalsy()
    });

    test('ipv6', async () => {
        const ipv6Regex = regex.ipv6Regex
        const match = ipv6Regex.exec('2001:db8:3333:4444:5555:6666:7777:8888')
        expect(match).toBeTruthy()

        const match2 = ipv6Regex.exec('2001:db8:3333:4444:5555:6666:7777:8888z')
        expect(match2).toBeFalsy()

        const match3 = ipv6Regex.exec('2001:db8:3333:4444:5555:6666:7777:888z')
        expect(match3).toBeFalsy()

        const match4 = ipv6Regex.exec('2001:db8:3333::8888')
        expect(match4).toBeTruthy()

        const match5 = ipv6Regex.exec('2001:db8:3333:4444:5555:8888')
        expect(match5).toBeFalsy()
    });

    test('torv3', async () => {
        const torv3Regex = regex.torv3Regex
        const match = torv3Regex.exec('gwdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqd.onion')
        expect(match).toBeTruthy()

        const match2 = torv3Regex.exec('gwdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqd.onion$')
        expect(match2).toBeFalsy()

        const match3 = torv3Regex.exec('wdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqd.onion')
        expect(match3).toBeFalsy()

        const match5 = torv3Regex.exec('gwdllz5g7vky2q4gr45zguvoajzf33czreca3a3exosftx72ekppkuqdonion')
        expect(match5).toBeFalsy()
    });

    test('domainName', async () => {
        const domainRegex = regex.domainNameRegex
        const match1 = domainRegex.exec('swiss-reign.ch')
        expect(match1).toBeTruthy()

        const match2 = domainRegex.exec('sub.swiss-reign.ch')
        expect(match2).toBeTruthy()

        const match3 = domainRegex.exec('.sub.swiss-reign.ch')
        expect(match3).toBeFalsy()

        const match4 = domainRegex.exec('localhost')
        expect(match4).toBeFalsy()

        const match5 = domainRegex.exec('myreallylongnamelongerthan64charactersmyreallylongnamelongerthan64charactersmyreallylongnamelongerthan64characters')
        expect(match5).toBeFalsy()

        const match6 = domainRegex.exec('sub.swiss-reign.ch-')
        expect(match6).toBeFalsy()

        const match7 = domainRegex.exec('127.0.0.1')
        expect(match7).toBeFalsy()
    });


    test('port', async () => {
        const portRegex = regex.portRegex
        const match = portRegex.exec('120')
        expect(match).toBeTruthy()

        const match2 = portRegex.exec('1200000')
        expect(match2).toBeFalsy()

        const match3 = portRegex.exec('a')
        expect(match3).toBeFalsy()

        const match5 = portRegex.exec('-1')
        expect(match5).toBeFalsy()
    });
});




src/regex.ts
/**
 * DO NOT use capture groups for regex here. They are not supported on all browsers!
 */

export const pubkeyRegex = /^(0[23][\dabcdef]{64})$/
export const ipv4Regex = /^(((25[0-5]|(2[0-4]|1\d|[1-9]|)\d)\.?\b){4})$/
export const ipv6Regex = /^((([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])))$/
export const torv3Regex = /^([\w\d]{56,56}\.onion)$/
export const domainNameRegex = /^(\b((?=[a-z0-9-]{1,63}\.)(xn--)?[a-z0-9]+(-[a-z0-9]+)*\.)+[a-z]{2,63}\b)$/
export const portRegex = /^(([1-9][0-9]{0,3}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5]))$/
export const squareBracketsAddressRegex = /^([\[](.*)[\]])(:?(\d{1,5}))?$/

src/ParseError.ts
export enum ParseFailureCode {
    /**
     * Host part is invalid.
     * Host can be either an IPv4, IPv6, Tor v3 address or a domain name.
     */
    INVALID_HOST = 'invalidHost',
    /**
     * Ipv6 is invalid.
     */
    INVALID_IPV6 = 'invalidIpv6',
    /**
     * Provided Ipv6 is ambigious and can not be separated from the port without guessing.
     */
    AMBIGUOUS_IPV6 = 'ambiguousIpv6',
    /**
     * Port is invalid. Must be between 1 and 65535.
     */
    INVALID_PORT = 'invalidPort',
    /**
     * Lightning pubkey is invalid.
     */
    INVALID_PUBKEY = 'invalidPubkey',
    /**
     * @ is missing or too many @ present.
     */
    INVALID_ATS = 'invalidAts',
}

export class ParseError extends Error {
    constructor(message: string, public code: ParseFailureCode) {
        super(message)
        this.name = 'ParseError'
    }
}

src/parse.ts
import { ParseError, ParseFailureCode } from "./ParseError"
import { domainNameRegex, ipv4Regex, ipv6Regex, portRegex, pubkeyRegex, squareBracketsAddressRegex, torv3Regex } from "./regex";

export type HostType = 'ipv6' | 'ipv4' | 'torv3' | 'domain';

export interface ParsedConnectionString {
    /**
     * Host part of the connection string. Could be an IPv4, IPv6, Tor v3 address or a domain name.
     */
    host: string, 
    /**
     * Port number of the connection string if provided. Is between 1 and 65535.
     */
    port?: number,
    /**
     * Lightning public key.
     */
    pubkey: string,
    /**
     * Host type.
     */
    hostType: HostType 
}


/**
 * Splits the address to a host and optionally a port. Throws ParseError if the address is invalid.
 * @param addressString 
 * @param opts.portMandatory Last number after colon is always port. This removes the IPv6 ambiguity because the port is always at the end.
 * @returns 
 */
export function splitHostAndPort(addressString: string, opts: Partial<{portMandatory: boolean}> = {}): {
    notValidatedHost: string,
    notValidatedPort?: string
} {
    // Easy, host is in square brackets [] and port is optionally behind a colon.
    const squareBracketMatch = squareBracketsAddressRegex.exec(addressString) 
    if (squareBracketMatch) {
        const squareBracketContent = squareBracketMatch[2]
        const squareBracketPort = squareBracketMatch[4]
        return {
            notValidatedHost: squareBracketContent || '',
            notValidatedPort: squareBracketPort
        }
    }

    // Max 1 colon. So it is not an ipv6.
    const colonCount = (addressString.split(":").length - 1)
    if (colonCount <= 1) {
        // Either tor or ipv4
        const splitted = addressString.split(":")
        return {
            notValidatedHost: splitted[0],
            notValidatedPort: splitted[1]
        }
    }

    if (opts.portMandatory) {
        // Port is provided in the address string. So we can split on the last colon.
        const splitted = addressString.split(":")
        const host = splitted.slice(0, -1).join(':')
        const port = splitted[splitted.length - 1]
        return {
            notValidatedHost: host,
            notValidatedPort: port
        }
    }

    // Ipv6 without square brackets []. Here the guessing begins.
    const isCompressed = addressString.includes('::')
    if (!isCompressed) {
        // Ipv6 consists of 8 groups plus an optional port.
        const groups = addressString.split(":")
        if (groups.length > 9 || groups.length < 8) {
            // Invalid ipv6
            throw new ParseError('Invalid host. Invalid ipv6?', ParseFailureCode.INVALID_HOST)
        }
        const host = groups.slice(0, 8).join(':')
        const port = groups[8]
        return {
            notValidatedHost: host,
            notValidatedPort: port
        }
    }

    // Ipv6 compressed
    // We can't determine the port here as it is ambigious.
    // 2001:db8::8888:9735 may as well be a IPv6 (2001:db8::8888:9735) OR a IPv6:port (2001:db8::8888 and port 9735).
    throw new ParseError('Compressed ipv6 host without square brackets []', ParseFailureCode.AMBIGUOUS_IPV6)
}

/**
 * Parses a host. Valid formats: ipv4, ipv6, torv3, domain. Throws ParseError if the host is invalid.
 * @param hostString 
 * @returns 
 */
export function parseHost(hostString: string): { host: string, hostType: HostType } {

    const ipv4Match = ipv4Regex.exec(hostString)
    if (ipv4Match) {
        return {
            host: hostString,
            hostType: 'ipv4'
        }
    }

    const ipv6Match = ipv6Regex.exec(hostString)
    if (ipv6Match) {
        return {
            host: hostString,
            hostType: 'ipv6'
        }
    }

    const torv3Match = torv3Regex.exec(hostString)
    if (torv3Match) {
        return {
            host: hostString,
            hostType: 'torv3'
        }
    }

    const domainMatch = domainNameRegex.exec(hostString)
    if (domainMatch) {
        return {
            host: hostString,
            hostType: 'domain'
        }
    }

    throw new ParseError('Invalid host.', ParseFailureCode.INVALID_HOST)
}

/**
 * Address consists of a host and an optional port. The host may be in square brackets [].
 * Examples: 127.0.0.1:9375, [::1]:9375, synonym.to:9375, 127.0.0.1.
 * Throws ParseError if the address is invalid.
 * @param addressString 
 * @param opts.portMandatory If true, throws ParseError if port is not provided.
 * @returns 
 */
export function parseAddress(addressString: string, opts: Partial<{portMandatory: boolean}> = {}): { host: string, port?: number, hostType: HostType } {
    const { notValidatedHost, notValidatedPort } = splitHostAndPort(addressString, {portMandatory: opts.portMandatory})
    const { host, hostType } = parseHost(notValidatedHost)

    const port = parsePort(notValidatedPort, {portMandatory: opts.portMandatory})

    return {
        port: port,
        host: host,
        hostType: hostType
    }
}

/**
 * Parse port number. Must be between 1 and 65535. Throws ParseError if the port is invalid.
 * @param portString 
 * @param opts.portMandatory If true, throws ParseError if port is not provided.
 * @returns 
 */
export function parsePort(portString?: string, opts: Partial<{portMandatory: boolean}> = {}): number | undefined {
    if (!portString) {
        if (opts.portMandatory) {
            throw new ParseError('Port is mandatory but is not provided.', ParseFailureCode.INVALID_PORT)
        } else {
            return undefined
        }
    }
    const portMatch = portRegex.exec(portString)
    if (!portMatch) {
        throw new ParseError('Invalid port.', ParseFailureCode.INVALID_PORT)
    }
    return Number.parseInt(portString)
}

/**
 * Parses a pubkey. Throws ParseError if the pubkey is invalid.
 * @param pubkeyString 
 * @returns 
 */
export function parsePubkey(pubkeyString: string): string {
    const pubkeyMatch = pubkeyRegex.exec(pubkeyString)
    if (!pubkeyMatch) {
        throw new ParseError('Invalid pubkey.', ParseFailureCode.INVALID_PUBKEY)
    }
    return pubkeyString
}


/**
 * Simple connection string parse. Throws ConnectionStringValidationError if the string is invalid.
 * @param connectionString Connection string to parse.
 * @param opts.portMandatory Last number after colon is always port. This removes the IPv6 ambiguity because the port is always at the end.
 * @returns 
 */
export function parseConnectionString(connectionString: string, opts: Partial<{portMandatory: boolean}> = {}): ParsedConnectionString {
    const split = connectionString.split('@')
    if (split.length !== 2) {
        throw new ParseError('@ does not split the string in two parts.', ParseFailureCode.INVALID_ATS)
    }
    const pubkeyString = split[0]

    const pubkey = parsePubkey(pubkeyString)

    const addressString = split[1]
    const address = parseAddress(addressString, {portMandatory: opts.portMandatory})

    return {
        pubkey: pubkey,
        hostType: address.hostType,
        host: address.host,
        port: address.port
    }
}